pkg/util/util.go

```diff
@@ -23,3 +25,82 @@ func GetCertPool(paths []string) (*x509.CertPool, error) {
 	}
 	return pool, nil
 }
+
+func IsEndpointAllowed(endpoint *url.URL, allowedDomains []string) bool {
+	hostname := endpoint.Hostname()
+
+	for _, allowedDomain := range allowedDomains {
+		allowedHost, allowedPort := SplitHostPort(allowedDomain)
+		if allowedHost == "" {
+			continue
+		}
+
+		if isHostnameBelongsToAllowedHost(hostname, allowedDomain) {
+			// the domain names match, now validate the ports
+			// if the whitelisted domain's port is '*', allow all ports
+			// if the whitelisted domain contains a specific port, only allow that port
+			// if the whitelisted domain doesn't contain a port at all, only allow empty redirect ports ie http and https
+			redirectPort := endpoint.Port()
+			if allowedPort == "*" ||
+				allowedPort == redirectPort ||
+				(allowedPort == "" && redirectPort == "") {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func isHostnameBelongsToAllowedHost(hostname, allowedHost string) bool {
+	// check if we have a perfect match between hostname and allowedHost
+	if hostname == strings.TrimPrefix(allowedHost, ".") ||
+		hostname == strings.TrimPrefix(allowedHost, "*.") {
+		return true
+	}
+
+	// check if hostname is a sub domain of the allowedHost
+	if (strings.HasPrefix(allowedHost, ".") && strings.HasSuffix(hostname, allowedHost)) ||
+		(strings.HasPrefix(allowedHost, "*.") && strings.HasSuffix(hostname, allowedHost[1:])) {
+		return true
+	}
+
+	return false
+}
+
+// SplitHostPort separates host and port. If the port is not valid, it returns
+// the entire input as host, and it doesn't check the validity of the host.
+// Unlike net.SplitHostPort, but per RFC 3986, it requires ports to be numeric.
+// *** taken from net/url, modified validOptionalPort() to accept ":*"
+func SplitHostPort(hostport string) (host, port string) {
```

Comment:
This can be private right?

```suggestion
func splitHostPort(hostport string) (host, port string) {
```

Reply:
Not possible since the function is needed in `validation` and `redirect` package.